### PR TITLE
feat(7411): JSX namespaced attribute syntax not supported

### LINKED ISSUE #7411: JSX namespaced attribute syntax not supported
**TypeScript Version:** 

1.8.2

**Background**:

I have been working on trying to get https://github.com/eslint/typescript-eslint-parser, a TypeScript parser plugin for ESLint, off the ground, and the job this weekend has been to begin adding JSX support.

ESLint uses espree and so the aim of the project is to convert the output of the tsc to an AST which espree expects. We already have a solid suite of JSX tests to develop against (taken from the espree project itself), but I have come up against a tsc error in one of them so far.

My issue is that it seems currently the tsc does not accept this "namespaced attribute" JSX syntax:

**Code**

``` ts
<a n:foo="bar"> {value} <b><c /></b></a>;

```

**Expected behavior:**
Please note the AST produced by espree (see in particular the `tokens` array): http://astexplorer.net/#/B46lew7I59

**Actual behavior:**
...compared to the one produced by the tsc (note the issues found in `parseDiagnostics`):
http://astexplorer.net/#/G5L3CptVNq

Removing the `n:` from `n:foo` resolves the parsing issues.

### LINKED ISSUE #47219: JSX: <svg : path /> parsed incorrectly
# Bug Report

### 🔎 Search Terms

jsx tag namespace

### 🕗 Version & Regression Information

v4.5.4, v4.6.0-dev.20211221

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about JSX

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.6.0-dev.20211221#code/JYWwDg9gTgLgBAJQKYEMDG8BmUIjgIilQ3wG4AoTAVwDsNgIa4BZATwGFdIakaYAKAJRwA3uThwiMKlCYAeAM4A3AOZwAXHDAoYACzgB6AHwUAvkA)

### 💻 Code

```ts
import React from "react";
function MyComponent() {
  return <svg : path />;
}
```

### 🙁 Actual behavior

TypeScript reports errors:
> Type 'true' is not assignable to type 'string | undefined'.
> Identifier expected.

TypeScript compiler generates undesired JavaScript:
```js
import React from "react";
function MyComponent() {
    return React.createElement("svg", { path: true });
}
```

It looks like TypeScript is interpreting `svg : path` similar to `svg path`. According to the [JSX specification](http://facebook.github.io/jsx/), whitespace is allowed around the `:` token.

### 🙂 Expected behavior

TypeScript treats the code as if it was the following:
```js
import React from "react";
function MyComponent() {
  return <svg:path />;  // Note: no spaces around ':'.
}
```
generating the following JavaScript:
```js
import React from "react";
function MyComponent() {
    return React.createElement("svg:path", null);
}
```

Babel, ESLint, and Flow interpret the code this way. ESBuild seems to have the same interpretation as TypeScript.

### PR DESCRIPTION
**Pull request by a-tarasyuk**
Fixes #7411
Fixes #47219

[JSX Spec](https://facebook.github.io/jsx/)

```
JSXElementName:
 - JSXIdentifier
 - JSXNamespacedName
 - JSXMemberExpression
```

```
JSXAttributeName:
 - JSXIdentifier
 - JSXNamespacedName
```

```
JSXNamespacedName:
 - JSXIdentifier : JSXIdentifier
```

### CODE REVIEW

**Review by weswigham — CHANGES_REQUESTED** [review]
Can we get tests with more variety of odd namespace names in both tag name and property name positions? Like `Component:foo` or `some-thing:bar`? Specifically, in the case of the first one, I'd like to know if it tries to somehow lookup a `Component:foo` local variable (which'd be odd) for the type of the tag. (You could technically make this with an assignment to `globalThis["Component:foo"]`, except the lookup for the tag probably has no shot of resolving at runtime - do namespaced custom tags even work?) which I'm down with just looking up intrinsics named `svg:path` and so on and props named `ns:prop` and so on, so that part of the checking seems fine. But do we need errors when a react-ish emit mode is set? `react` doesn't use namespaces as far as I'm aware. Maybe this should be flagged in some way because of that? (it's flagged in babel afaik)

Additionally, ping @RyanCavanaugh for the go/no go on this; we didn't want em for a long time, but I guess some framework (vue2?) uses the namespace syntax now, even though you have to silence an error in babel to use it.

**Comment by a-tarasyuk** [comment]
@weswigham Thanks for the review. 

> Can we get tests with more variety of odd namespace names in both tag name and property name positions? Like Component:foo or some-thing:bar?

I'm not sure I fully understood the requested changes. Do you mean something like this? Or do you mean that I need to define global types to check name resolution?

```ts
// @jsx: preserve
// @filename: a.tsx

<A:foo></A:foo>;
```

Could you provide some examples that I can use as tests?

**Comment by weswigham** [comment]
> Do you mean something like this?

Yeah, with both capital-letter-starting and dash-containing names. (The first being an indication of function components and the later indicating non-identifier builtins). For checking intrinsics, adding namespaced dashed names to intrinsic attributes should be fine. For the capital-letter variant, I don't think there's any reasonable way to check it or even transpile it, after all `React.createElement(Component:foo)` is obviously invalid syntax, so it probably needs to be an error.

**Comment by a-tarasyuk** [comment]
> For the capital-letter variant, I don't think there's any reasonable way to check it or even transpile it, after all React.createElement(Component:foo) is obviously invalid syntax, so it probably needs to be an error.

Shouldn't we transpile it as [`React.createElement("Component:foo")`](https://github.com/microsoft/TypeScript/pull/47356/files#diff-8d310bea1fa9287ac4aca0a8e45488c91b51214c8cff027964c21a4cb40f340dR11)?

**Comment by weswigham** [comment]
No - only intrinsic names do that, and only names starting with a lower case character are intrinsic names.

**Comment by a-tarasyuk** [comment]
@weswigham Should TS parse it as `JSXNamespacedName` and throw an error in the _checker_? Or at the `parser` stage?

**Comment by weswigham** [comment]
Mmmm, checker is better for error recovery, usually.

**Comment by a-tarasyuk** [comment]
> Additionally, ping @RyanCavanaugh for the go/no go on this; we didn't want em for a long time, but I guess some framework (vue2?) uses the namespace syntax now, even though you have to silence an error in babel to use it.

@RyanCavanaugh What do you think about [JSXNamespacedName](https://facebook.github.io/jsx/#prod-JSXNamespacedName) support?

**Review by weswigham — CHANGES_REQUESTED** [review]
Still waiting for @RyanCavanaugh to say if we actually wanna support this like this, or if we were only looking to fix the weird parser error on `tag : name`.

**Review by weswigham — APPROVED** [review]
This looks OK to me, but I'm going to kick it over to @RyanCavanaugh for final approval and merging after community tests have come back, since it's up to him if this much support for namespace names is actually what we wanted (or if we just wanted better parser recovery).

**Comment by weswigham** [comment]
@typescript-bot test this
@typescript-bot run dt
@typescript-bot test top100
@typescript-bot perf test this
@typescript-bot pack this

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -47821,25 +47824,11 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
...
-            if (check) {
-                return check;
-            }
+        if (isPropertyAccessExpression(node) && isJsxNamespacedName(node.expression)) {
+            return grammarErrorOnNode(node.expression, Diagnostics.JSX_property_access_expressions_cannot_include_JSX_namespace_names);
         }
-
-        function checkGrammarJsxNestedIdentifier(name: MemberName | ThisExpression) {
-            if (isIdentifier(name) && idText(name).indexOf(":") !== -1) {
-                return grammarErrorOnNode(name, Diagnostics.JSX_property_access_expressions_cannot_include_JSX_namespace_names);
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts:47841]: Does this need to be removed from `diagnosticMessages.json` now?

File: src/compiler/diagnosticMessages.json [hunk 2]
```diff
@@ -2893,6 +2893,10 @@
         "category": "Error",
         "code": 2638
     },
+    "React components cannot include JSX namespace names": {
```
Comment (DanielRosenwasser) [inline src/compiler/diagnosticMessages.json:2896]: This is specific to attributes? If so:

```suggestion
    "React components cannot support namespaced attribute names.": {
```

File: src/compiler/checker.ts (hunk 1 above)
Comment (a-tarasyuk) [reply src/compiler/checker.ts:47841]: This message is used to handle the following cases

```ts
<b:c.x></b:c.x>;
   ~~~
!!! error TS2633: JSX property access expressions cannot include JSX namespace names
```

```md
JSXNamespacedName:
 - JSXIdentifier : JSXIdentifier
```
https://facebook.github.io/jsx/#prod-JSXNamespacedName

**Review by DanielRosenwasser — APPROVED** [review]

**Comment by eps1lon** [comment]
It's not really clear to me whether this was targetted just at attributes or also element names. The spec does allow namespaced names as far as I can tell so TypeScript should allow it. And considering `@types/fbt` has been out for a while using namespaced element names, it seems to have worked in the past. But maybe how it worked in the past was not intended?

Either way, can somebody take a look at the test added in https://github.com/microsoft/TypeScript/pull/53784/files since that should resolve the element type to `fbt:enum` and not treat `fbt` and `enum` as different types. Bisected with the test added in https://github.com/microsoft/TypeScript/pull/53784 and this PR is the first one to break that test.

**Pull request merged**